File: steganography.py
from PIL import Image
import base64
import os
from cryptography.fernet import Fernet
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC

# The delimiter signals the end of the hidden message
DELIMITER = "====EOF===="
EOF_MARKER = b"====STEALTHGUARD===="

def generate_key(password: str, salt: bytes = b'cybersecurity_salt') -> bytes:
    """Generates a cryptographic key from a password using PBKDF2."""
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=32,
        salt=salt,
        iterations=100000,
    )
    key = base64.urlsafe_b64encode(kdf.derive(password.encode()))
    return key

def encrypt_message(message: str, password: str) -> str:
    """Encrypts a string message using AES (Fernet) and a password."""
    key = generate_key(password)
    f = Fernet(key)
    encrypted_bytes = f.encrypt(message.encode())
    # Return as base64 string and append delimiter so it can be found during extraction
    return encrypted_bytes.decode('utf-8') + DELIMITER

def decrypt_message(encrypted_str: str, password: str) -> str:
    """Decrypts a message using the password."""
    key = generate_key(password)
    f = Fernet(key)
    try:
        # Delimiter should already be stripped by decode_image, but just in case
        if encrypted_str.endswith(DELIMITER):
            encrypted_str = encrypted_str[:-len(DELIMITER)]
            
        decrypted_bytes = f.decrypt(encrypted_str.encode())
        return decrypted_bytes.decode('utf-8')
    except Exception as e:
        # Decryption failed (wrong password or corrupted data)
        raise ValueError("Invalid password or corrupted data.")

def text_to_binary(text: str) -> str:
    """Converts a string to a sequence of 8-bit binary numbers."""
    return ''.join(format(ord(char), '08b') for char in text)

def binary_to_text(binary_str: str) -> str:
    """Converts a sequence of 8-bit binary numbers back to a string."""
    text = ""
    for i in range(0, len(binary_str), 8):
        byte = binary_str[i:i+8]
        if len(byte) == 8:
            text += chr(int(byte, 2))
    return text

def encode_image(image_path: str, message: str, password: str, output_path: str) -> bool:
    """Hides an encrypted message inside an image without changing its file size."""
    # 1. Encrypt the message
    encrypted_msg = encrypt_message(message, password)
    
    # 2. Read the original image bytes
    with open(image_path, "rb") as f:
        img_bytes = f.read()
        
    # 3. Append the marker and encrypted message to the end of the image file
    # This prevents the file size from blowing up, keeping it exactly the same as original (+ a few bytes)
    with open(output_path, "wb") as f:
        f.write(img_bytes)
        f.write(EOF_MARKER)
        f.write(encrypted_msg.encode('utf-8'))
        
    return True

def decode_image(image_path: str, password: str) -> str:
    """Extracts and decrypts a hidden message from an image."""
    # First, try the EOF method which preserves file size
    with open(image_path, "rb") as f:
        img_bytes = f.read()
        
    marker_index = img_bytes.find(EOF_MARKER)
    if marker_index != -1:
        encrypted_bytes = img_bytes[marker_index + len(EOF_MARKER):]
        try:
            extracted_text = encrypted_bytes.decode('utf-8')
            if extracted_text.endswith(DELIMITER):
                encrypted_msg = extracted_text.split(DELIMITER)[0]
                return decrypt_message(encrypted_msg, password)
        except Exception:
            pass # Fallback to LSB if EOF extraction fails

    # --- FALLBACK TO OLD LSB METHOD ---
    # 1. Read image
    try:
        img = Image.open(image_path)
        if img.mode != 'RGB':
            img = img.convert('RGB')
    except Exception as e:
        raise FileNotFoundError(f"Could not open or find the image {image_path}: {e}")
        
    width, height = img.size
    pixels = img.load()
    
    binary_data = ""
    extracted_text = ""
    
    for y in range(height):
        for x in range(width):
            pixel = pixels[x, y]
            
            # Extract LSB from R, G, B
            for i in range(3):
                binary_data += str(pixel[i] & 1)
                
                # Check every 8 bits
                if len(binary_data) >= 8 and len(binary_data) % 8 == 0:
                    byte = binary_data[-8:]
                    extracted_text += chr(int(byte, 2))
                    
                    # Check if we reached the delimiter
                    if extracted_text.endswith(DELIMITER):
                        encrypted_msg = extracted_text.split(DELIMITER)[0]
                        return decrypt_message(encrypted_msg, password)
                        
    # If delimiter isn't found, it might not be a valid stego image
    raise ValueError("No hidden message found in this image.")

File: Steganography.kt
package com.example.stealthguard // You can change this to match your Android project's package

import android.graphics.BitmapFactory
import android.graphics.Color
import android.util.Base64
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.nio.ByteBuffer
import java.security.MessageDigest
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

object Steganography {

    private const val DELIMITER = "====EOF===="
    private val EOF_MARKER = "====STEALTHGUARD====".toByteArray(Charsets.UTF_8)
    private val SALT = "cybersecurity_salt".toByteArray(Charsets.UTF_8)

    /**
     * Generates a 32-byte key from password and salt using PBKDF2HMAC-SHA256
     * with 100,000 iterations, matching Python cryptography.hazmat's PBKDF2HMAC.
     */
    private fun generateRawKey(password: String): ByteArray {
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        val spec = PBEKeySpec(password.toCharArray(), SALT, 100000, 256)
        return factory.generateSecret(spec).encoded
    }

    /**
     * Encrypts a string message into a Python `cryptography.fernet` compatible token.
     */
    fun encryptMessage(message: String, password: String): String {
        val rawKey = generateRawKey(password)
        val signingKey = rawKey.copyOfRange(0, 16)
        val encryptionKey = rawKey.copyOfRange(16, 32)

        val secureRandom = SecureRandom()
        val iv = ByteArray(16)
        secureRandom.nextBytes(iv)

        val cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(encryptionKey, "AES"), IvParameterSpec(iv))
        val ciphertext = cipher.doFinal(message.toByteArray(Charsets.UTF_8))

        val timestamp = System.currentTimeMillis() / 1000L
        val version = 0x80.toByte()

        val dataToSign = ByteArrayOutputStream().apply {
            write(version.toInt())
            write(ByteBuffer.allocate(8).putLong(timestamp).array())
            write(iv)
            write(ciphertext)
        }.toByteArray()

        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(signingKey, "HmacSHA256"))
        val hmac = mac.doFinal(dataToSign)

        val fernetToken = ByteArrayOutputStream().apply {
            write(dataToSign)
            write(hmac)
        }.toByteArray()

        // Base64 URL-safe encoding without padding (matches Python's base64.urlsafe_b64encode)
        // Note: Python's fernet strips padding or expects standard URL safe base64, 
        // Android's NO_PADDING ensures compatibility.
        val tokenString = Base64.encodeToString(fernetToken, Base64.URL_SAFE or Base64.NO_WRAP)
        
        // Append delimiter like the Python backend
        return tokenString + DELIMITER
    }

    /**
     * Decrypts a Fernet token string.
     */
    fun decryptMessage(encryptedStr: String, password: String): String {
        var tokenStr = encryptedStr
        if (tokenStr.endsWith(DELIMITER)) {
            tokenStr = tokenStr.substring(0, tokenStr.length - DELIMITER.length)
        }

        val fernetToken = Base64.decode(tokenStr, Base64.URL_SAFE or Base64.NO_WRAP)
        
        if (fernetToken.size < 1 + 8 + 16 + 32) {
            throw IllegalArgumentException("Invalid token size")
        }

        val version = fernetToken[0]
        if (version != 0x80.toByte()) {
            throw IllegalArgumentException("Invalid version")
        }

        val iv = fernetToken.copyOfRange(9, 25)
        val ciphertext = fernetToken.copyOfRange(25, fernetToken.size - 32)
        val hmac = fernetToken.copyOfRange(fernetToken.size - 32, fernetToken.size)

        val dataToSign = fernetToken.copyOfRange(0, fernetToken.size - 32)

        val rawKey = generateRawKey(password)
        val signingKey = rawKey.copyOfRange(0, 16)
        val encryptionKey = rawKey.copyOfRange(16, 32)

        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(signingKey, "HmacSHA256"))
        val computedHmac = mac.doFinal(dataToSign)

        if (!MessageDigest.isEqual(hmac, computedHmac)) {
            throw IllegalArgumentException("Invalid password or corrupted data (signature mismatch).")
        }

        val cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(encryptionKey, "AES"), IvParameterSpec(iv))
        val decryptedText = cipher.doFinal(ciphertext)

        return String(decryptedText, Charsets.UTF_8)
    }

    /**
     * Hides an encrypted message inside an image or audio file by appending to EOF.
     * Keeps file size practically the same.
     */
    fun encodeMedia(mediaPath: File, message: String, password: String, outputPath: File): Boolean {
        val encryptedMsg = encryptMessage(message, password)
        
        FileInputStream(mediaPath).use { inputStream ->
            FileOutputStream(outputPath).use { outputStream ->
                inputStream.copyTo(outputStream)
                outputStream.write(EOF_MARKER)
                outputStream.write(encryptedMsg.toByteArray(Charsets.UTF_8))
            }
        }
        return true
    }

    /**
     * Extracts and decrypts a hidden message from an image or audio file.
     */
    fun decodeMedia(mediaPath: File, password: String): String {
        val mediaBytes = mediaPath.readBytes()
        val markerIndex = indexOf(mediaBytes, EOF_MARKER)

        if (markerIndex != -1) {
            val encryptedBytes = mediaBytes.copyOfRange(markerIndex + EOF_MARKER.size, mediaBytes.size)
            try {
                val extractedText = String(encryptedBytes, Charsets.UTF_8)
                if (extractedText.contains(DELIMITER)) {
                    val encryptedMsg = extractedText.substringBefore(DELIMITER) + DELIMITER
                    return decryptMessage(encryptedMsg, password)
                }
            } catch (e: Exception) {
                // Fallback to LSB if EOF extraction fails
            }
        }

        // --- FALLBACK TO OLD LSB METHOD (IMAGES ONLY) ---
        return decodeMediaLsb(mediaPath, password)
    }

    /**
     * Helper function to find subarray index
     */
    private fun indexOf(array: ByteArray, target: ByteArray): Int {
        if (target.isEmpty()) return 0
        outer@ for (i in 0 until array.size - target.size + 1) {
            for (j in target.indices) {
                if (array[i + j] != target[j]) {
                    continue@outer
                }
            }
            return i
        }
        return -1
    }

    /**
     * Decodes a hidden message using the LSB method with Android's BitmapFactory.
     */
    private fun decodeMediaLsb(mediaPath: File, password: String): String {
        // BitmapFactory is the Android equivalent of PIL.Image
        val bitmap = BitmapFactory.decodeFile(mediaPath.absolutePath)
            ?: throw IllegalArgumentException("No hidden message found in this media file.")

        val width = bitmap.width
        val height = bitmap.height

        val binaryData = StringBuilder()
        val extractedText = StringBuilder()

        for (y in 0 until height) {
            for (x in 0 until width) {
                val pixel = bitmap.getPixel(x, y)
                
                // Extract RGB values
                val rgb = intArrayOf(Color.red(pixel), Color.green(pixel), Color.blue(pixel))
                
                // Extract LSB from R, G, B
                for (i in 0..2) {
                    binaryData.append(rgb[i] and 1)
                    
                    // Check every 8 bits
                    if (binaryData.length >= 8 && binaryData.length % 8 == 0) {
                        val byteStr = binaryData.substring(binaryData.length - 8)
                        extractedText.append(Integer.parseInt(byteStr, 2).toChar())
                        
                        // Check if we reached the delimiter
                        if (extractedText.toString().endsWith(DELIMITER)) {
                            val encryptedMsg = extractedText.toString().substringBefore(DELIMITER) + DELIMITER
                            return decryptMessage(encryptedMsg, password)
                        }
                    }
                }
            }
        }

        throw IllegalArgumentException("No hidden message found in this image.")
    }
}
